crates/sui-storage/src/indexes.rs

```diff
@@ -238,21 +257,85 @@ impl IndexStore {
+        // Objects in `deleted_owners` must be owned by `Owner::Address` befoer the tx,
+        // hence must appear in the tx inputs.
+        // They also mut be coin type (see `AuthorityState::commit_certificate`).
+        let coin_delete_keys = object_index_changes
+            .deleted_owners
+            .iter()
+            .filter_map(|(owner, obj_id)| {
+                // If it's not in `input_coins`, then it's not a coin type. Skip.
+                let coin_type_tag = input_coins.get(obj_id)?
+                .coin_type_maybe().unwrap_or_else(|| {
+                    panic!(
+                        "object_id: {:?} in input_coins is not a coin type, input_coins: {:?}, tx_digest: {:?}",
+                        obj_id, input_coins, digest
+                    )
+                });
+                Some((*owner, coin_type_tag.to_string(), *obj_id))
+            }).collect::<Vec<_>>();
+        trace!(
+            tx_digset=?digest,
+            "coin_delete_keys: {:?}",
+            coin_delete_keys,
+        );
+
+        batch.delete_batch(&self.tables.coin_index, coin_delete_keys.into_iter())?;
+
+        // 2. Upsert new owner, by looking at `object_index_changes.new_owners`.
+        // For a object to appear in `new_owners`, it must be owned by `Owner::Address` after the tx.
+        // It also must not be deleted, hence appear in written_coins (see `AuthorityState::commit_certificate`)
+        // It also must be a coin type (see `AuthorityState::commit_certificate`).
+        // Here the coin could be transfered to a new address, to simply have the metadata changed (digest, balance etc)
+        // due to a successful or failed transaction.
+        let coin_add_keys = object_index_changes
+        .new_owners
+        .iter()
+        .filter_map(|((owner, obj_id), obj_info)| {
+            // TODO assert_eq obj_ref with obj_info?
+            // If it's in written_coins, then it's not a coin. Skip it.
+            let (_obj_ref, obj, _write_kind) = written_coins.get(obj_id)?;
+            let coin_type_tag = obj.coin_type_maybe().unwrap_or_else(|| {
+                panic!(
+                    "object_id: {:?} in written_coins is not a coin type, written_coins: {:?}, tx_digest: {:?}",
+                    obj_id, written_coins, digest
+                )
+            });
+            let coin = obj.as_coin_maybe().unwrap_or_else(|| {
+                panic!(
+                    "object_id: {:?} in written_coins cannot be deserialzied as a Coin, written_coins: {:?}, tx_digest: {:?}",
+                    obj_id, written_coins, digest
+                )
+            });
+            Some(((*owner, coin_type_tag.to_string(), *obj_id), (CoinInfo {version: obj_info.version, digest: obj_info.digest, balance: coin.balance.value(), previous_transaction: *digest})))
+        }).collect::<Vec<_>>();
+        trace!(
+            tx_digset=?digest,
+            "coin_add_keys: {:?}",
+            coin_add_keys,
+        );
```

Comment:
@longbowlu Could we've just did an `objects` table lookup to read the balance of new coin instead of passing this along like this. I may be naive in my thinking but seems like `post processing` is guaranteed to be blocking so that key is guaranteed to be there in `objects` table and we don't need to change anything in `commit certificate` to pass values and overall things stay much simple

Reply:
this is explained in the description, copying it here:

Why do we need this? Two reasons:
a. if an object is deleted, we can no longer get the content from the object table after the commit, then no way to know its coin type at that point. But to delete an entry in the coin index, we need to know its coin type.
b. it's much cheaper to grab them here compared to fetching them from the db later, especially when the number of coins is big which is the problem we tried to avoid initially.